src/haz3lcore/dynamics/TypeAssignment.re

```diff
@@ -0,0 +1,301 @@
+    List.fold_left((acc, t) => {acc && Typ.eq(t, ty)}, true, l)
+      ? Some(ty) : None
+  };
+};
+
+let delta_ty = (id: MetaVar.t, m: Statics.Map.t): option(Typ.t) => {
+  switch (Id.Map.find_opt(id, m)) {
+  | Some(InfoExp({mode, ctx, _})) =>
+    switch (mode) {
+    | Syn
+    | SynFun => Some(Unknown(Internal))
+    | Ana(ana_ty) => Some(Typ.normalize(ctx, ana_ty))
+    }
+  | _ => None
+  };
+};
+
+let ground = (ty: Typ.t): bool => {
+  switch (Transition.CastHelpers.ground_cases_of(ty)) {
+  | Ground => true
+  | _ => false
+  };
+};
+
+let dhpat_extend_ctx = (dhpat: DHPat.t, ty: Typ.t, ctx: Ctx.t): option(Ctx.t) => {
+  let rec dhpat_var_entry =
+          (dhpat: DHPat.t, ty: Typ.t): option(list(Ctx.entry)) => {
+    switch (dhpat) {
+    | Var(name) =>
+      let entry = Ctx.VarEntry({name, id: Id.invalid, typ: ty});
+      Some([entry]);
+    | Tuple(l1) =>
+      switch (ty) {
+      | Prod(l2) when List.length(l1) == List.length(l2) =>
+        let* l =
+          List.map2((dhp, typ) => {dhpat_var_entry(dhp, typ)}, l1, l2)
+          |> OptUtil.sequence;
+        Some(List.concat(l));
+      | _ => None
+      }
+    | Cons(dhp1, dhp2) =>
+      switch (ty) {
+      | List(typ) =>
+        let* l1 = dhpat_var_entry(dhp1, typ);
+        let* l2 = dhpat_var_entry(dhp2, List(typ));
+        Some(l1 @ l2);
+      | _ => None
+      }
+    | ListLit(typ1, l) =>
+      switch (ty) {
+      | List(typ2) when Typ.eq(typ1, typ2) =>
+        let* l =
+          List.map(dhp => {dhpat_var_entry(dhp, typ1)}, l)
+          |> OptUtil.sequence;
+        Some(List.concat(l));
+      | _ => None
+      }
+    | Ap(Constructor(_, typ), dhp) =>
+      let (ty1, ty2) = Typ.matched_arrow(ctx, typ);
+      Typ.eq(ty2, ty) ? dhpat_var_entry(dhp, ty1) : None;
+    | EmptyHole(_)
+    | NonEmptyHole(_)
+    | Wild
+    | ExpandingKeyword(_)
+    | InvalidText(_)
+    | BadConstructor(_)
+    | IntLit(_)
+    | FloatLit(_)
+    | BoolLit(_)
+    | StringLit(_)
+    | Constructor(_)
+    | Ap(_) => Some([])
+    };
+  };
+  let+ l = dhpat_var_entry(dhpat, ty);
+  List.fold_left((ctx, entry) => Ctx.extend(ctx, entry), ctx, l);
+};
+
+let rec typ_of_dhexp =
+        (ctx: Ctx.t, m: Statics.Map.t, dh: DHExp.t): option(Typ.t) => {
+  switch (dh) {
+  | EmptyHole(id, _) => delta_ty(id, m)
+  | NonEmptyHole(_, id, _, d) =>
+    switch (typ_of_dhexp(ctx, m, d)) {
+    | None => None
+    | Some(_) => delta_ty(id, m)
+    }
+  | FreeVar(id, _, _) => delta_ty(id, m)
+  | ExpandingKeyword(_)
+  | InvalidText(_) => Some(Unknown(Internal))
+  | InconsistentBranches(_, _, Case(d_scrut, d_rules, _)) =>
+    let* ty' = typ_of_dhexp(ctx, m, d_scrut);
+    let typ_cases =
+      d_rules
+      |> List.map((DHExp.Rule(dhp, de)) => {
+           let* ctx = dhpat_extend_ctx(dhp, ty', ctx);
+           typ_of_dhexp(ctx, m, de);
+         })
+      |> OptUtil.sequence;
+
+    switch (typ_cases) {
+    | None => None
+    | Some(_) => Some(Typ.Unknown(Internal))
```

Comment:
let's actually check to make sure that there is at least one branch with an inconsistent type here